pkg/pool-manager/pool.go

```diff
@@ -315,6 +315,101 @@ func GetMacPoolSize(rangeStart, rangeEnd net.HardwareAddr) (int64, error) {
 }
 
+func (p *PoolManager) createOrUpdateMacEntryInMacPoolMap(macAddress, instanceFullName, macInstanceKey string) {
+	p.macPoolMap[macAddress] = macEntry{
+		instanceName:         instanceFullName,
+		macInstanceKey:       macInstanceKey,
+		transactionTimestamp: noPendingTransaction,
+	}
+}
+
+// getInstanceMacMap creates a subset map from macPoolMap, holding only macs that belongs to a specific instance (pod/vm)
+func (p *PoolManager) getInstanceMacMap(instanceName string) (map[string]macEntry, error) {
+	instanceMacMap := map[string]macEntry{}
+
+	for macAddress, macEntry := range p.macPoolMap {
+		if macEntry.instanceName == instanceName {
+			instanceMacMap[macAddress] = macEntry
+		}
+	}
+	return instanceMacMap, nil
+}
+
+// isMacUpdateRequired checks mac entry if an update is needed
+func (p *PoolManager) isMacUpdateRequired(macAddress string) bool {
+	if p.macPoolMap[macAddress].transactionTimestamp != noPendingTransaction {
+		return true
+	}
+	return false
+}
+
+// clearMacTransactionFromMacEntry clears mac entry's transactionTimestamp, signalling that no further update is needed
+func (p *PoolManager) clearMacTransactionFromMacEntry(macAddress string) {
+	macEntry := p.macPoolMap[macAddress]
+	macEntry.transactionTimestamp = noPendingTransaction
+	p.macPoolMap[macAddress] = macEntry
+}
+
+// removeMacEntry deletes a macEntry from macPoolMap
+func (p *PoolManager) removeMacEntry(macAddress string) {
+	delete(p.macPoolMap, macAddress)
+}
+
+// isMacReadyForTransactionUpdate checks latest vm persisted transaction TS passed the mac's transaction TS,
+// signalling that an macPoolMap update may be needed for this entry.
+func (p *PoolManager) isMacReadyForTransactionUpdate(macAddress, lastPersistentTransactionTimestamp string) (bool, error) {
+	macTransactionTime, err := time.Parse(time.RFC3339Nano, p.macPoolMap[macAddress].transactionTimestamp)
+	if err != nil {
+		log.Error(err, "failed to parse mac transaction time")
+		return false, errors.Wrapf(err, "mac address %s, timestamp %s", macAddress, p.macPoolMap[macAddress].transactionTimestamp)
+	}
+	lastPersistentTime, err := time.Parse(time.RFC3339Nano, lastPersistentTransactionTimestamp)
+	if err != nil {
+		log.Error(err, "failed to parse last persistent transaction time")
+		return false, errors.Wrapf(err, "mac address %s, timestamp %s", macAddress, lastPersistentTransactionTimestamp)
+	}
+
+	if macTransactionTime.After(lastPersistentTime) {
+		return false, nil
+	}
+	return true, nil
+}
+
+// isMacTransactionStale checks if mac entry has not been handled for more than waitTime, setting it as stale.
+func (p *PoolManager) isMacTransactionStale(macAddress string) (bool, error) {
+	macTransactionTime, err := time.Parse(time.RFC3339Nano, p.macPoolMap[macAddress].transactionTimestamp)
+	if err != nil {
+		log.Error(err, "failed to parse mac transaction time")
+		return false, err
+	}
+
+	if p.now().After(macTransactionTime.Add(time.Duration(p.waitTime) * time.Second)) {
```

Comment:
>Why is not waitTime already a Duration ?

because it's pulled from the container params , so it's parsed as int. but I can change if you feel strongly about it :) 

fixed the second comment for better readability, thanks!

Reply:
Try to convert p.waitTime at the same place that you load it, so we check problems at the load place.